### ISSUE TITLE
Bug: Can't escape / (forward slash) in no-restricted-syntax rule.

### ISSUE THREAD

**Issue by RoyRao2333** [eslint/eslint#16555]
### Environment

Node version: 16.16.0
pnpm version: 7.15.0
Local ESLint version: 8.23.0
Operating System: macOS 13.0.1, M1 Pro

### What parser are you using?

@typescript-eslint/parser

### What did you do?

Forward slashes are not escaped in no-restricted-syntax.
Regex as below:

```typescript
const extMatchingRex = /^(?:[\S\t]*[-.& ,+!@#$%\^*\(\);\/\|\<\>"\'?=:_\[\]\{\}~`\t])?(jpg|jpeg|png|gif|svg|webp|apng|pdf|doc|docx|ppt|pptx|xls|xlsx|txt|rtf|md|html|xml|zip|rar|7z|iso|exe|rmvb|avi|mpeg|mp4|m4v|mov|asf|flv|f4v|3gp|mp3|wma|wav|ape|flac|ogg|aac|m4a)(?:[-.& ,+!@#$%\^*\(\);\/\|\<\>"\'?=:_\[\]\{\}~`\t][\S\t]*)?$/
```

Usage as below:

```typescript
{
    selector: `Literal[value=${extMatchingRex}]`,
    message: 'Use it uppercased.',
},
```

No luck putting my regex in a string or RegExp either.

### What did you expect to happen?

It should match the following strings:

```typescript
const jpg = 'jpg';
const jpeg = '.jpeg';
const mixed_png = 'This is a hh .png. hh file';
```

### What actually happened?

<img width="672" alt="image" src="https://user-images.githubusercontent.com/31413093/202106348-c7cc8973-a1f7-49af-a18c-793e4aa98f63.png">

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

As you can see, my forward slashed is well escaped, but it came up with the error above.
After I removed ```\/```, it went well... How to fix this?

**Comment by mdjermanovic** [comment]
Hi @RoyRao2333, thanks for the issue!

This seems to be caused by a problem in esquery, explained here:

 https://eslint.org/docs/latest/developer-guide/selectors#known-issues

Can you try with `\u002F` instead of `\/` and check if the selector works as expected after the change?

```js
const extMatchingRex = /^(?:[\S\t]*[-.& ,+!@#$%\^*\(\);\u002F\|\<\>"\'?=:_\[\]\{\}~`\t])?(jpg|jpeg|png|gif|svg|webp|apng|pdf|doc|docx|ppt|pptx|xls|xlsx|txt|rtf|md|html|xml|zip|rar|7z|iso|exe|rmvb|avi|mpeg|mp4|m4v|mov|asf|flv|f4v|3gp|mp3|wma|wav|ape|flac|ogg|aac|m4a)(?:[-.& ,+!@#$%\^*\(\);\u002F\|\<\>"\'?=:_\[\]\{\}~`\t][\S\t]*)?$/
```

**Comment by RoyRao2333** [comment]
> 

Yes! It works like a charm 🎉 Can this be fixed someday later? Thanks for the big help!

**Comment by RoyRao2333** [comment]
> Hi @RoyRao2333, thanks for the issue!
> 
> This seems to be caused by a problem in esquery, explained here:
> 
> https://eslint.org/docs/latest/developer-guide/selectors#known-issues
> 
> Can you try with `\u002F` instead of `\/` and check if the selector works as expected after the change?
> 
> ```js
> const extMatchingRex = /^(?:[\S\t]*[-.& ,+!@#$%\^*\(\);\u002F\|\<\>"\'?=:_\[\]\{\}~`\t])?(jpg|jpeg|png|gif|svg|webp|apng|pdf|doc|docx|ppt|pptx|xls|xlsx|txt|rtf|md|html|xml|zip|rar|7z|iso|exe|rmvb|avi|mpeg|mp4|m4v|mov|asf|flv|f4v|3gp|mp3|wma|wav|ape|flac|ogg|aac|m4a)(?:[-.& ,+!@#$%\^*\(\);\u002F\|\<\>"\'?=:_\[\]\{\}~`\t][\S\t]*)?$/
> ```

Yes! It works like a charm 🎉 Can this be fixed someday later? Thanks for the big help!

**Comment by mdjermanovic** [comment]
This can only be fixed in esquery. I'll leave this issue open as a reminder for us to update the esquery dependency when https://github.com/estools/esquery/issues/68 is fixed.

**Comment by RoyRao2333** [comment]
> This can only be fixed in esquery. I'll leave this issue open as a reminder for us to update the esquery dependency when https://github.com/estools/esquery/issues/68 is fixed.

Got it. Thanks for the help! 🎉

**Comment by gfyoung** [comment]
Just curious, do we actually see a path forward to addressing that upstream issue?

That repository hasn't been updated for almost two years.  Perhaps we need to cross-ping some folks over there?

**Comment by JoshuaKGoldberg** [comment]
For posterity, this is the basic test case I'm using in `tests/lib/rules/no-restricted-syntax.js` to test https://github.com/estools/esquery/pull/149:

```js
{
	code: `import values from 'some/path';`,
	options: ["ImportDeclaration[source.value=/^some\\/path$/]"],
	errors: [
		{
			messageId: "restrictedSyntax",
			data: {
				message:
					"Using 'ImportDeclaration[source.value=/^some\\/path$/]' is not allowed.",
			},
		},
	],
}
```
